### PR TITLE
[FIXED JENKINS-40228] - Remove Enable Security checkbox in the Global Security configuration

### PR DESCRIPTION
**Pull request by ksenia-nenasheva**
Follow up to PR https://github.com/jenkinsci/jenkins/pull/2900
See [JENKINS-40228](https://issues.jenkins-ci.org/browse/JENKINS-40228).

**Details:**
 - Checkbox _Enable security_ was deleted (Security is enabled by default from Jenkins 2.0)
 - Add _Security Realm :: None_

**Before:**

**After:**

### Proposed changelog entries

* Entry 1: -  Remove ES checkbox in the Global Security configuration
* ...

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### CODE REVIEW

**Comment by ksenia-nenasheva** [comment]
@jenkinsci/code-reviewers

**Comment by oleg-nenashev** [comment]
@jenkinsci/code-reviewers ping

**Comment by MarkEWaite** [comment]
I have a [Jenkins docker image](https://github.com/MarkEWaite/docker-lfs/tree/lts-with-plugins) based on current LTS and current plugins which intentionally does not enable security.  I use that as the basis for several different test configurations.  Those later test configurations have security enabled, but this configuration does not.

Won't this removal of the "Enable security" checkbox prevent me from enabling security on that docker instance?

File: core/src/main/java/hudson/security/SecurityRealm.java [hunk 1]
```diff
@@ -580,6 +573,21 @@ public Filter createFilter(FilterConfig filterConfig) {
         private Object readResolve() {
             return NO_AUTHENTICATION;
         }
+        
+        @Extension
```
Comment (daniel-beck) [inline core/src/main/java/hudson/security/SecurityRealm.java]: Similar to the "No SCM" option in jobs this should set ordinal to always appear last in the list.

File: core/src/main/java/hudson/security/GlobalSecurityConfiguration.java [hunk 2]
```diff
@@ -111,14 +111,10 @@ public boolean configure(StaplerRequest req, JSONObject json) throws hudson.mode
...
-            JSONObject security = json.getJSONObject("useSecurity");
-            j.setDisableRememberMe(security.optBoolean("disableRememberMe", false));
-            j.setSecurityRealm(SecurityRealm.all().newInstanceFromRadioList(security, "realm"));
-            j.setAuthorizationStrategy(AuthorizationStrategy.all().newInstanceFromRadioList(security, "authorization"));    
-        } else {
-            j.disableSecurity();
-        }
+
+        j.setDisableRememberMe(json.optBoolean("disableRememberMe", false));
+        j.setSecurityRealm(SecurityRealm.all().newInstanceFromRadioList(json, "realm"));
```
Comment (daniel-beck) [inline core/src/main/java/hudson/security/GlobalSecurityConfiguration.java]: It seems easy to lock oneself out by choosing None security realm, and almost any authorization strategy. Maybe we can check whether the security realm is set to the none realm, and if so, call `disableSecurity()`?

Form validation dependent on both fields would also be nice, if that's possible (don't remember right now).

**Review by daniel-beck — APPROVED** [review]
Was really surprised to see this unmerged, I actually thought this was already in, I consider this such an obvious improvement.

File: core/src/main/java/hudson/security/SecurityRealm.java (hunk 1 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/security/SecurityRealm.java]: +1

File: core/src/main/java/hudson/security/GlobalSecurityConfiguration.java (hunk 2 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/security/GlobalSecurityConfiguration.java]: Validation of radio-selectors is PITA though doable.

I agree that additional check would be useful. Maybe `Jenkins.getAuthorizationStrategy()` should just ignore the setting when the security realm is not set. And them we could hook on the existing Security Administrative Monitor

**Comment by oleg-nenashev** [comment]
> Was really surprised to see this unmerged

me too :)

**Review by oleg-nenashev — APPROVED** [review]
I addressed the merge conflict and the ordinal comment from @daniel-beck 
@jenkinsci/core I believe it can be integrated now

**Comment by oleg-nenashev** [comment]
So the ATH test will legitimately fail after the merge. It is going to be a chicken-and-egg, because we need to fix the test in ATH which depends on the core itself. So the only proper approach would be to have the ATH logic to behave differently depending on the version.

screenshot 2019-02-01 at 13 54 34


@raul-arabaolaza @olivergondza @MRamonLeon what would be your take on that?

**Review by Wadeck — APPROVED** [review]

**Comment by daniel-beck** [comment]
I think `on-hold` better represents the need for ATH compatibility based on @oleg-nenashev's latest comment.

**Comment by oleg-nenashev** [comment]
yes. It is a chicken-and-egg problem. There is a PR to ATH:
https://github.com/jenkinsci/acceptance-test-harness/pull/393 . AFAICT we
need ATH to work for both pre-change and post-change versions somehow

**Comment by batmat** [comment]
IMO Oleg is summarizing it well, I agree ATH change should adapt to both versions of Jenkins. 
I think the whole PR could be made green without needing to break`master` at any point.  

I would recommend this way:
1. Someone (not necessarily Ksenia) adjust https://github.com/jenkinsci/acceptance-test-harness/pull/393 to be compatible with both versions of Jenkins (before and with this change in Jenkins Core)
2. Bump in this PR the reference of the SHA version of ATH to use in `essentials.yaml`: https://github.com/jenkinsci/jenkins/blob/9330635821cbcb430b4223b471692f2c9ae862db/essentials.yml#L4 (even using the SHA for ATH PR should work I think)

Then
3. the PR build goes green here
4. we merge on ATH side the PR
5. we bump here to ATH new `master`
6. We merge here if still green

Phew. 

---

Before all this, I think we should have a PR that just attempts to bump to latest ATH `master` state. So the work above doesn't end up handling some unrelated error.

WDYT about this proposal @olivergondza ?

**Comment by batmat** [comment]
http://github.com/jenkinsci/jenkins/pull/3990 will most likely be a prerequisite for any move here. (I.e. close the gap of the sha we're currently using for ATH and the current ATH master branch state)

**Comment by olivergondza** [comment]
@batmat, yes that is the way to go (note sure of the top of my head if the ATH PR would work or we need a feature branch as well). Though https://github.com/jenkinsci/acceptance-test-harness/pull/393 does not address the change implemented here, it is meant for different core PR altogether.

**Comment by jglick** [comment]
> It is going to be a chicken-and-egg, because we need to fix the test in ATH which depends on the core itself.

This is why I have long advocated for ATH to have a particular version of Jenkins it tests by default.

File: core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy [hunk 3]
```diff
@@ -25,14 +25,21 @@ l.layout(norefresh:true, permission:app.ADMINISTER, title:my.displayName, csscla
...
 
-                f.entry(title:_("Access Control")) {
+                f.entry(title:_("Security Realm")) {
                     table(style:"width:100%") {
                         f.descriptorRadioList(title:_("Security Realm"),varName:"realm",         instance:app.securityRealm,         descriptors:h.filterDescriptors(app, SecurityRealm.all()))
+                    }
+                }
+            }
+
+            f.section(title:_("Autorization")) {
```
Comment (jglick) [inline core/src/main/resources/hudson/security/GlobalSecurityConfiguration/index.groovy]:
```suggestion
            f.section(title:_("Authorization")) {
```

**Review by jglick — APPROVED** [review]
Thank you! Have been waiting for this simple cleanup for years.

Regarding ATH, you could just patch `essentials.yml#ath/athRevision` here, no?

**Comment by oleg-nenashev** [comment]
I will try to get the ATH over the line next week

**Comment by MRamonLeon** [comment]
@ksenia-nenasheva could you please resolve the conflicting file and add the [suggested change from Jesse](https://github.com/jenkinsci/jenkins/pull/2930#discussion_r279561116)?

Very appreciated, thank you.

**Comment by ksenia-nenasheva** [comment]
@jglick @MRamonLeon 
Thank you! Sorry for the delay.

**Comment by jglick** [comment]
Looks like there is still a merge conflict here?

**Review by varyvol — APPROVED** [review]

**Comment by varyvol** [comment]
Changes look good but @oleg-nenashev did the ATH changes landed?

**Review by alecharp — APPROVED** [review]
@oleg-nenashev you mentioned that you would try to land some ATH related stuff, is it included?

**Comment by timja** [comment]
> I am on it

Next steps here?
Does master need merging in?

**Comment by varyvol** [comment]
@ksenia-nenasheva @oleg-nenashev it seems the ATH changes were merged. Is there anything else pending?

**Comment by jglick** [comment]
`TriggerRemoteBuildsTest.triggerBuildRemotely` is failing.

**Pull request merged**